library/default-synths-extra.scd

```diff
@@ -523,3 +523,109 @@ SynthDef(\superreese,  {|out, pan, freq, sustain, accelerate, detune=0, voice=0|
 	OffsetOut.ar(out, DirtPan.ar(sound, ~dirt.numChannels, pan, env));
 }).add;
 );
+
+
+// 6-op FM synth (DX7-like)
+//
+// Works a bit different from the original DX7.  Instead of algorithms, you set the amount
+// of modulation every operator receives from other operators and itself (feedback), virtually
+// providing an endless number of possible combinations (algorithms).
+//
+// responds to
+//   lfofreq (pitch modulation frequency)
+//   lfodepth (pitch modulation amplitude)
+//
+// Each operator responds to
+//   amp (audible amplitude)
+//   ratio (frequency ratio)
+//   detune (in Hz)
+//   eglevel[1-4] (4 envelope generator levels)
+//   egrate[1-4] (4 envelope generator rates)
+//
+// The syntax for operator arguments is <argumentName + opIndex>[modulatorIndex | egIndex]
+//
+// For example:
+// amp1 1      (op1 as carrier with full amplitude)
+// ratio2 2.3  (op2 frequency ratio)
+// mod11 0.5   (op1 feedback)
+// mod12 0.78  (op1 modulation amount by op2)
+// detune1 0.2 (op1 detune)
+// eglevel12   (op1 EG level2)
+// egrate11    (op1 EG rate1) -- higher values go FASTER!
+
+(
+SynthDef(\superfm, {
+	var sustain = \sustain.kr(1);
```

Comment:
if your controls are controlled one-per-event it would be much more efficient to use `.ir` instead of `.kr`.

Reply:
I'm not sure how events are treated in SuperDirt.  Is an event triggered every time a Tidal message is recieved?  How do `wave` arguments in Tidal translate to SuperDirt?  I tried to follow the other SynthDefs structure, and since `rate` is not specified their arguments default to `.kr`, don't they?

Reply:
>I'm not sure how events are treated in SuperDirt. Is an event triggered every time a Tidal message is recieved?

Yes

> How do `wave` arguments in Tidal translate to SuperDirt?

they are just passed on without translation, and received by the synth if it defines the argument. In principle, you could also send lists/arrays, but this is not yet implemented in the tidal-sclang communication.

> I tried to follow the other SynthDefs structure, and since rate is not specified their arguments default to .kr, don't they?

the SynthDefs have a list of rates specified in the end (`[\ir, \ir, \ir]` e.g.), this is how this is normally done. But you can conveniently specify this in your SynthDef, where you define arguments in the style `symbol.kr(defaultValue)`, or `symbol.ir(defaultValue)`.

Reply:
sure.  I was thinking it might be good to change some of them to `\ir` but then realized that it's nice to be able to modulate all the parameters.  What are your thoughts?

Reply:
Yes, you could leave it as it is, maybe you can then map controls in supercollider to continuous processes. Only if you want to have the most efficient implementation for tidal, then you'd use initial rate, because it saves electricity.